pkg/quota/evaluator/core/persistent_volume_claims.go

```diff
@@ -23,63 +23,124 @@ import (
-func PersistentVolumeClaimUsageFunc(object runtime.Object) api.ResourceList {
-	pvc, ok := object.(*api.PersistentVolumeClaim)
-	if !ok {
-		return api.ResourceList{}
-	}
-	result := api.ResourceList{}
-	result[api.ResourcePersistentVolumeClaims] = resource.MustParse("1")
-	if request, found := pvc.Spec.Resources.Requests[api.ResourceStorage]; found {
-		result[api.ResourceRequestsStorage] = request
+// makeMatchedResourceNamesFunc returns a function that returns the list of resources matched
+func makeMatchedResourceNamesFunc(storageClassStore cache.Store) generic.MatchesResourceNamesFunc {
+	allResources := []api.ResourceName{api.ResourcePersistentVolumeClaims, api.ResourceRequestsStorage}
+	return func() []api.ResourceName {
+		resourceNames := []api.ResourceName{}
+		for _, c := range storageClassStore.List() {
+			if storageClass, ok := c.(*storage.StorageClass); ok {
+				prefix := storageClass.Name
+				for _, resource := range allResources {
+					resourceName := api.ResourceName(prefix + storageClassSuffix + string(resource))
+					resourceNames = append(resourceNames, resourceName)
+				}
+			}
+		}
+		fmt.Printf("quota: matchesfunc: %v\n", resourceNames)
+		return resourceNames
 	}
-	return result
 }
 
-// PersistentVolumeClaimConstraintsFunc verifies that all required resources are present on the claim
-// In addition, it validates that the resources are valid (i.e. requests < limits)
-func PersistentVolumeClaimConstraintsFunc(required []api.ResourceName, object runtime.Object) error {
-	pvc, ok := object.(*api.PersistentVolumeClaim)
-	if !ok {
-		return fmt.Errorf("unexpected input object %v", object)
-	}
+// makePersistentVolumeClaimUsageFunc returns a function that knows how to measure usage for pvcs
+func makePersistentVolumeClaimUsageFunc(storageClassStore cache.Store, matchesResourceNameFunc generic.MatchesResourceNamesFunc) generic.UsageFunc {
+	return func(object runtime.Object) api.ResourceList {
+		pvc, ok := object.(*api.PersistentVolumeClaim)
+		if !ok {
+			return api.ResourceList{}
+		}
+
+		result := api.ResourceList{}
+		matchedResources := matchesResourceNameFunc()
+		for _, matchedResource := range matchedResources {
+			result[matchedResource] = resource.MustParse("0")
+		}
 
-	requiredSet := quota.ToSet(required)
-	missingSet := sets.NewString()
-	pvcUsage := PersistentVolumeClaimUsageFunc(pvc)
-	pvcSet := quota.ToSet(quota.ResourceNames(pvcUsage))
-	if diff := requiredSet.Difference(pvcSet); len(diff) > 0 {
-		missingSet.Insert(diff.List()...)
+		result[api.ResourcePersistentVolumeClaims] = resource.MustParse("1")
+		if request, found := pvc.Spec.Resources.Requests[api.ResourceStorage]; found {
+			result[api.ResourceRequestsStorage] = request
+			// charge usage to the storage class (if present)
+			if storageClassRef, storageClassFound := pvc.Annotations["volume.beta.kubernetes.io/storage-class"]; storageClassFound {
```

Comment:
My props for ACLs would need to have an admission controller that would check whether or not the user can ask for that storage class. It should deny the user if they don't have authorization for the requested class (exists or not)